// @elastic/eui/lib/services/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Random = exports.noOverflowPlacement = exports.Comparators = exports.SortableProperties = exports.SortDirection = exports.SortDirectionType = exports.PropertySortType = exports.getSecureRelForTarget = exports.checkHrefAndOnClick = exports.Pager = exports.formatText = exports.formatNumber = exports.formatDate = exports.formatBoolean = exports.formatAuto = exports.VISUALIZATION_COLORS = exports.isColorDark = exports.CENTER_ALIGNMENT = exports.RIGHT_ALIGNMENT = exports.LEFT_ALIGNMENT = exports.htmlIdGenerator = exports.comboBoxKeyCodes = exports.cascadingMenuKeyCodes = exports.accessibleClickKeys = exports.keyCodes = undefined;

var _accessibility = require('./accessibility');

Object.defineProperty(exports, 'accessibleClickKeys', {
  enumerable: true,
  get: function get() {
    return _accessibility.accessibleClickKeys;
  }
});
Object.defineProperty(exports, 'cascadingMenuKeyCodes', {
  enumerable: true,
  get: function get() {
    return _accessibility.cascadingMenuKeyCodes;
  }
});
Object.defineProperty(exports, 'comboBoxKeyCodes', {
  enumerable: true,
  get: function get() {
    return _accessibility.comboBoxKeyCodes;
  }
});
Object.defineProperty(exports, 'htmlIdGenerator', {
  enumerable: true,
  get: function get() {
    return _accessibility.htmlIdGenerator;
  }
});

var _alignment = require('./alignment');

Object.defineProperty(exports, 'LEFT_ALIGNMENT', {
  enumerable: true,
  get: function get() {
    return _alignment.LEFT_ALIGNMENT;
  }
});
Object.defineProperty(exports, 'RIGHT_ALIGNMENT', {
  enumerable: true,
  get: function get() {
    return _alignment.RIGHT_ALIGNMENT;
  }
});
Object.defineProperty(exports, 'CENTER_ALIGNMENT', {
  enumerable: true,
  get: function get() {
    return _alignment.CENTER_ALIGNMENT;
  }
});

var _color = require('./color');

Object.defineProperty(exports, 'isColorDark', {
  enumerable: true,
  get: function get() {
    return _color.isColorDark;
  }
});
Object.defineProperty(exports, 'VISUALIZATION_COLORS', {
  enumerable: true,
  get: function get() {
    return _color.VISUALIZATION_COLORS;
  }
});

var _format = require('./format');

Object.defineProperty(exports, 'formatAuto', {
  enumerable: true,
  get: function get() {
    return _format.formatAuto;
  }
});
Object.defineProperty(exports, 'formatBoolean', {
  enumerable: true,
  get: function get() {
    return _format.formatBoolean;
  }
});
Object.defineProperty(exports, 'formatDate', {
  enumerable: true,
  get: function get() {
    return _format.formatDate;
  }
});
Object.defineProperty(exports, 'formatNumber', {
  enumerable: true,
  get: function get() {
    return _format.formatNumber;
  }
});
Object.defineProperty(exports, 'formatText', {
  enumerable: true,
  get: function get() {
    return _format.formatText;
  }
});

var _paging = require('./paging');

Object.defineProperty(exports, 'Pager', {
  enumerable: true,
  get: function get() {
    return _paging.Pager;
  }
});

var _prop_types = require('./prop_types');

Object.defineProperty(exports, 'checkHrefAndOnClick', {
  enumerable: true,
  get: function get() {
    return _prop_types.checkHrefAndOnClick;
  }
});

var _security = require('./security');

Object.defineProperty(exports, 'getSecureRelForTarget', {
  enumerable: true,
  get: function get() {
    return _security.getSecureRelForTarget;
  }
});

var _sort = require('./sort');

Object.defineProperty(exports, 'PropertySortType', {
  enumerable: true,
  get: function get() {
    return _sort.PropertySortType;
  }
});
Object.defineProperty(exports, 'SortDirectionType', {
  enumerable: true,
  get: function get() {
    return _sort.SortDirectionType;
  }
});
Object.defineProperty(exports, 'SortDirection', {
  enumerable: true,
  get: function get() {
    return _sort.SortDirection;
  }
});
Object.defineProperty(exports, 'SortableProperties', {
  enumerable: true,
  get: function get() {
    return _sort.SortableProperties;
  }
});
Object.defineProperty(exports, 'Comparators', {
  enumerable: true,
  get: function get() {
    return _sort.Comparators;
  }
});

var _overflow = require('./overflow');

Object.defineProperty(exports, 'noOverflowPlacement', {
  enumerable: true,
  get: function get() {
    return _overflow.noOverflowPlacement;
  }
});

var _random = require('./random');

Object.defineProperty(exports, 'Random', {
  enumerable: true,
  get: function get() {
    return _random.Random;
  }
});

var _key_codes = require('./key_codes');

var keyCodes = _interopRequireWildcard(_key_codes);

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key]; } } newObj.default = obj; return newObj; } }

exports.keyCodes = keyCodes; // Export all keyCodes under a `keyCodes` named variable
