// @elastic/eui/lib/components/stat/stat.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.EuiStat = exports.ALIGNMENTS = exports.COLORS = void 0;

require("core-js/modules/es6.object.assign");

require("core-js/modules/web.dom.iterable");

require("core-js/modules/es6.array.iterator");

require("core-js/modules/es6.object.keys");

var _react = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _classnames = _interopRequireDefault(require("classnames"));

var _text = require("../text");

var _title = require("../title/title");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = Object.defineProperty && Object.getOwnPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : {}; if (desc.get || desc.set) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } } newObj.default = obj; return newObj; } }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

var colorToClassNameMap = {
  default: null,
  subdued: 'euiStat__title--dark',
  primary: 'euiStat__title--primary',
  secondary: 'euiStat__title--secondary',
  danger: 'euiStat__title--danger',
  accent: 'euiStat__title--accent'
};
var COLORS = Object.keys(colorToClassNameMap);
exports.COLORS = COLORS;
var textAlignToClassNameMap = {
  left: 'euiStat--leftAligned',
  center: 'euiStat--centerAligned',
  right: 'euiStat--rightAligned'
};
var ALIGNMENTS = Object.keys(textAlignToClassNameMap);
exports.ALIGNMENTS = ALIGNMENTS;

var EuiStat = function EuiStat(_ref) {
  var children = _ref.children,
      className = _ref.className,
      description = _ref.description,
      title = _ref.title,
      titleSize = _ref.titleSize,
      titleColor = _ref.titleColor,
      textAlign = _ref.textAlign,
      reverse = _ref.reverse,
      rest = _objectWithoutProperties(_ref, ["children", "className", "description", "title", "titleSize", "titleColor", "textAlign", "reverse"]);

  var classes = (0, _classnames.default)('euiStat', textAlignToClassNameMap[textAlign], className);
  var titleClasses = (0, _classnames.default)('euiStat__title', colorToClassNameMap[titleColor]);

  var descriptionDisplay = _react.default.createElement(_text.EuiText, {
    size: "s",
    className: "euiStat__description"
  }, _react.default.createElement("p", null, description));

  var titleDisplay = _react.default.createElement(_title.EuiTitle, {
    size: titleSize,
    className: titleClasses
  }, _react.default.createElement("p", null, title));

  var statDisplay;

  if (reverse) {
    statDisplay = _react.default.createElement(_react.Fragment, null, titleDisplay, descriptionDisplay);
  } else {
    statDisplay = _react.default.createElement(_react.Fragment, null, descriptionDisplay, titleDisplay);
  }

  return _react.default.createElement("div", _extends({
    className: classes
  }, rest), statDisplay, children);
};

exports.EuiStat = EuiStat;
EuiStat.propTypes = {
  /**
   * Set the title (value) text
   */
  title: _propTypes.default.node.isRequired,

  /**
   * Set the description (label) text
   */
  description: _propTypes.default.node.isRequired,

  /**
   * Places the title (value) above the description (label)
   */
  reverse: _propTypes.default.bool.isRequired,

  /**
   * Define the size of the title text. See EuiTitle for sizing options ('s', 'm', 'l'... etc)
   */
  titleSize: _propTypes.default.oneOf(_title.TITLE_SIZES),

  /**
   * Define the color of the title text
   */
  titleColor: _propTypes.default.oneOf(COLORS),

  /**
   * Define how you want the content aligned
   */
  textAlign: _propTypes.default.oneOf(ALIGNMENTS),

  /**
   * Appends additional classes to parent
   */
  className: _propTypes.default.string,

  /**
   * Additional content that appears after the title and description
   */
  children: _propTypes.default.node
};
EuiStat.defaultProps = {
  titleColor: 'default',
  textAlign: 'left',
  titleSize: 'l',
  reverse: false
};
EuiStat.__docgenInfo = {
  "description": "",
  "methods": [],
  "displayName": "EuiStat",
  "props": {
    "titleColor": {
      "defaultValue": {
        "value": "'default'",
        "computed": false
      },
      "type": {
        "name": "enum",
        "value": [{
          "value": "\"default\"",
          "computed": false
        }, {
          "value": "\"subdued\"",
          "computed": false
        }, {
          "value": "\"primary\"",
          "computed": false
        }, {
          "value": "\"secondary\"",
          "computed": false
        }, {
          "value": "\"danger\"",
          "computed": false
        }, {
          "value": "\"accent\"",
          "computed": false
        }]
      },
      "required": false,
      "description": "Define the color of the title text"
    },
    "textAlign": {
      "defaultValue": {
        "value": "'left'",
        "computed": false
      },
      "type": {
        "name": "enum",
        "value": [{
          "value": "\"left\"",
          "computed": false
        }, {
          "value": "\"center\"",
          "computed": false
        }, {
          "value": "\"right\"",
          "computed": false
        }]
      },
      "required": false,
      "description": "Define how you want the content aligned"
    },
    "titleSize": {
      "defaultValue": {
        "value": "'l'",
        "computed": false
      },
      "type": {
        "name": "enum",
        "computed": true,
        "value": "TITLE_SIZES"
      },
      "required": false,
      "description": "Define the size of the title text. See EuiTitle for sizing options ('s', 'm', 'l'... etc)"
    },
    "reverse": {
      "defaultValue": {
        "value": "false",
        "computed": false
      },
      "type": {
        "name": "bool"
      },
      "required": false,
      "description": "Places the title (value) above the description (label)"
    },
    "title": {
      "type": {
        "name": "node"
      },
      "required": true,
      "description": "Set the title (value) text"
    },
    "description": {
      "type": {
        "name": "node"
      },
      "required": true,
      "description": "Set the description (label) text"
    },
    "className": {
      "type": {
        "name": "string"
      },
      "required": false,
      "description": "Appends additional classes to parent"
    },
    "children": {
      "type": {
        "name": "node"
      },
      "required": false,
      "description": "Additional content that appears after the title and description"
    }
  }
};
