// @elastic/eui/lib/components/search_bar/query/default_syntax.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.defaultSyntax = undefined;

var _ast = require('./ast');

var _predicate = require('../../../services/predicate');

function _toConsumableArray(arr) { if (Array.isArray(arr)) { for (var i = 0, arr2 = Array(arr.length); i < arr.length; i++) { arr2[i] = arr[i]; } return arr2; } else { return Array.from(arr); } }

// eslint-disable-line import/no-unresolved

var unescapeValue = function unescapeValue(value) {
  return value.replace(/\\([:\-\\])/, '$1');
};

var escapeValue = function escapeValue(value) {
  return value.replace(/([:\-\\])/, '\\$1');
};

var parser = /*
              * Generated by PEG.js 0.10.0.
              *
              * http://pegjs.org/
              */function () {
  "use strict";
  function peg$subclass(child, parent) {
    function ctor() {
      this.constructor = child;
    }ctor.prototype = parent.prototype;child.prototype = new ctor();
  }function peg$SyntaxError(message, expected, found, location) {
    this.message = message;this.expected = expected;this.found = found;this.location = location;this.name = "SyntaxError";if (typeof Error.captureStackTrace === "function") {
      Error.captureStackTrace(this, peg$SyntaxError);
    }
  }peg$subclass(peg$SyntaxError, Error);peg$SyntaxError.buildMessage = function (expected, found) {
    var DESCRIBE_EXPECTATION_FNS = { literal: function literal(expectation) {
        return "\"" + literalEscape(expectation.text) + "\"";
      }, "class": function _class(expectation) {
        var escapedParts = "",
            i;for (i = 0; i < expectation.parts.length; i++) {
          escapedParts += expectation.parts[i] instanceof Array ? classEscape(expectation.parts[i][0]) + "-" + classEscape(expectation.parts[i][1]) : classEscape(expectation.parts[i]);
        }return "[" + (expectation.inverted ? "^" : "") + escapedParts + "]";
      }, any: function any(expectation) {
        return "any character";
      }, end: function end(expectation) {
        return "end of input";
      }, other: function other(expectation) {
        return expectation.description;
      } };function hex(ch) {
      return ch.charCodeAt(0).toString(16).toUpperCase();
    }function literalEscape(s) {
      return s.replace(/\\/g, '\\\\').replace(/"/g, '\\"').replace(/\0/g, '\\0').replace(/\t/g, '\\t').replace(/\n/g, '\\n').replace(/\r/g, '\\r').replace(/[\x00-\x0F]/g, function (ch) {
        return '\\x0' + hex(ch);
      }).replace(/[\x10-\x1F\x7F-\x9F]/g, function (ch) {
        return '\\x' + hex(ch);
      });
    }function classEscape(s) {
      return s.replace(/\\/g, '\\\\').replace(/\]/g, '\\]').replace(/\^/g, '\\^').replace(/-/g, '\\-').replace(/\0/g, '\\0').replace(/\t/g, '\\t').replace(/\n/g, '\\n').replace(/\r/g, '\\r').replace(/[\x00-\x0F]/g, function (ch) {
        return '\\x0' + hex(ch);
      }).replace(/[\x10-\x1F\x7F-\x9F]/g, function (ch) {
        return '\\x' + hex(ch);
      });
    }function describeExpectation(expectation) {
      return DESCRIBE_EXPECTATION_FNS[expectation.type](expectation);
    }function describeExpected(expected) {
      var descriptions = new Array(expected.length),
          i,
          j;for (i = 0; i < expected.length; i++) {
        descriptions[i] = describeExpectation(expected[i]);
      }
      descriptions.sort();if (descriptions.length > 0) {
        for (i = 1, j = 1; i < descriptions.length; i++) {
          if (descriptions[i - 1] !== descriptions[i]) {
            descriptions[j] = descriptions[i];j++;
          }
        }descriptions.length = j;
      }switch (descriptions.length) {case 1:
          return descriptions[0];case 2:
          return descriptions[0] + " or " + descriptions[1];default:
          return descriptions.slice(0, -1).join(", ") + ", or " + descriptions[descriptions.length - 1];}
    }function describeFound(found) {
      return found ? "\"" + literalEscape(found) + "\"" : "end of input";
    }return "Expected " + describeExpected(expected) + " but " + describeFound(found) + " found.";
  };function peg$parse(input, options) {
    options = options !== void 0 ? options : {};var peg$FAILED = {},
        peg$startRuleFunctions = { Query: peg$parseQuery },
        peg$startRuleFunction = peg$parseQuery,
        peg$c0 = function peg$c0(clauses) {
      return clauses;
    },
        peg$c1 = function peg$c1() {
      return [];
    },
        peg$c2 = function peg$c2(head, clause) {
      return clause;
    },
        peg$c3 = function peg$c3(head, tail) {
      return [head].concat(_toConsumableArray(tail));
    },
        peg$c4 = "-",
        peg$c5 = peg$literalExpectation("-", false),
        peg$c6 = function peg$c6(value) {
      return AST.Term.mustNot(value);
    },
        peg$c7 = function peg$c7(value) {
      return AST.Term.must(value);
    },
        peg$c8 = function peg$c8(value) {
      return AST.Is.mustNot(value);
    },
        peg$c9 = function peg$c9(value) {
      return AST.Is.must(value);
    },
        peg$c10 = "is:",
        peg$c11 = peg$literalExpectation("is:", false),
        peg$c12 = function peg$c12(value) {
      return value;
    },
        peg$c13 = function peg$c13(fv) {
      return AST.Field.mustNot(fv.field, fv.value);
    },
        peg$c14 = function peg$c14(fv) {
      return AST.Field.must(fv.field, fv.value);
    },
        peg$c15 = ":",
        peg$c16 = peg$literalExpectation(":", false),
        peg$c17 = function peg$c17(field, value) {
      return { field: field, value: value };
    },
        peg$c18 = peg$otherExpectation("field name"),
        peg$c19 = function peg$c19() {
      return unescapeValue(text());
    },
        peg$c20 = peg$otherExpectation("field value"),
        peg$c21 = "(",
        peg$c22 = peg$literalExpectation("(", false),
        peg$c23 = /^[oO]/,
        peg$c24 = peg$classExpectation(["o", "O"], false, false),
        peg$c25 = /^[rR]/,
        peg$c26 = peg$classExpectation(["r", "R"], false, false),
        peg$c27 = function peg$c27(head, value) {
      return value;
    },
        peg$c28 = ")",
        peg$c29 = peg$literalExpectation(")", false),
        peg$c30 = function peg$c30(head, tail) {
      return [head].concat(_toConsumableArray(tail));
    },
        peg$c31 = peg$otherExpectation("term"),
        peg$c32 = "\"",
        peg$c33 = peg$literalExpectation("\"", false),
        peg$c34 = function peg$c34(phrase) {
      return phrase;
    },
        peg$c35 = "\\",
        peg$c36 = peg$literalExpectation("\\", false),
        peg$c37 = /^[:\-\\]/,
        peg$c38 = peg$classExpectation([":", "-", "\\"], false, false),
        peg$c39 = peg$otherExpectation("alpha numeric"),
        peg$c40 = /^[a-zA-Z0-9]/,
        peg$c41 = peg$classExpectation([["a", "z"], ["A", "Z"], ["0", "9"]], false, false),
        peg$c42 = peg$otherExpectation("whitespace"),
        peg$c43 = /^[ \t\n\r]/,
        peg$c44 = peg$classExpectation([" ", "\t", "\n", "\r"], false, false),
        peg$currPos = 0,
        peg$savedPos = 0,
        peg$posDetailsCache = [{ line: 1, column: 1 }],
        peg$maxFailPos = 0,
        peg$maxFailExpected = [],
        peg$silentFails = 0,
        peg$result;if ("startRule" in options) {
      if (!(options.startRule in peg$startRuleFunctions)) {
        throw new Error("Can't start parsing from rule \"" + options.startRule + "\".");
      }peg$startRuleFunction = peg$startRuleFunctions[options.startRule];
    }function text() {
      return input.substring(peg$savedPos, peg$currPos);
    }function location() {
      return peg$computeLocation(peg$savedPos, peg$currPos);
    }function expected(description, location) {
      location = location !== void 0 ? location : peg$computeLocation(peg$savedPos, peg$currPos);throw peg$buildStructuredError([peg$otherExpectation(description)], input.substring(peg$savedPos, peg$currPos), location);
    }function error(message, location) {
      location = location !== void 0 ? location : peg$computeLocation(peg$savedPos, peg$currPos);throw peg$buildSimpleError(message, location);
    }function peg$literalExpectation(text, ignoreCase) {
      return { type: "literal", text: text, ignoreCase: ignoreCase };
    }function peg$classExpectation(parts, inverted, ignoreCase) {
      return { type: "class", parts: parts, inverted: inverted, ignoreCase: ignoreCase };
    }function peg$anyExpectation() {
      return { type: "any" };
    }function peg$endExpectation() {
      return { type: "end" };
    }function peg$otherExpectation(description) {
      return { type: "other", description: description };
    }function peg$computePosDetails(pos) {
      var details = peg$posDetailsCache[pos],
          p;if (details) {
        return details;
      } else {
        p = pos - 1;while (!peg$posDetailsCache[p]) {
          p--;
        }details = peg$posDetailsCache[p];details = { line: details.line, column: details.column };while (p < pos) {
          if (input.charCodeAt(p) === 10) {
            details.line++;details.column = 1;
          } else {
            details.column++;
          }p++;
        }peg$posDetailsCache[pos] = details;return details;
      }
    }function peg$computeLocation(startPos, endPos) {
      var startPosDetails = peg$computePosDetails(startPos),
          endPosDetails = peg$computePosDetails(endPos);return { start: { offset: startPos, line: startPosDetails.line, column: startPosDetails.column }, end: { offset: endPos, line: endPosDetails.line, column: endPosDetails.column } };
    }function peg$fail(expected) {
      if (peg$currPos < peg$maxFailPos) {
        return;
      }if (peg$currPos > peg$maxFailPos) {
        peg$maxFailPos = peg$currPos;peg$maxFailExpected = [];
      }peg$maxFailExpected.push(expected);
    }function peg$buildSimpleError(message, location) {
      return new peg$SyntaxError(message, null, null, location);
    }function peg$buildStructuredError(expected, found, location) {
      return new peg$SyntaxError(peg$SyntaxError.buildMessage(expected, found), expected, found, location);
    }function peg$parseQuery() {
      var s0, s1;s0 = peg$currPos;s1 = peg$parseClauses();if (s1 !== peg$FAILED) {
        peg$savedPos = s0;s1 = peg$c0(s1);
      }s0 = s1;if (s0 === peg$FAILED) {
        s0 = peg$currPos;s1 = peg$parsespace();if (s1 === peg$FAILED) {
          s1 = null;
        }if (s1 !== peg$FAILED) {
          peg$savedPos = s0;s1 = peg$c1();
        }s0 = s1;
      }return s0;
    }function peg$parseClauses() {
      var s0, s1, s2, s3, s4, s5, s6;s0 = peg$currPos;s1 = peg$parsespace();if (s1 === peg$FAILED) {
        s1 = null;
      }if (s1 !== peg$FAILED) {
        s2 = peg$parseClause();if (s2 !== peg$FAILED) {
          s3 = [];s4 = peg$currPos;s5 = peg$parsespace();if (s5 !== peg$FAILED) {
            s6 = peg$parseClause();if (s6 !== peg$FAILED) {
              peg$savedPos = s4;s5 = peg$c2(s2, s6);s4 = s5;
            } else {
              peg$currPos = s4;s4 = peg$FAILED;
            }
          } else {
            peg$currPos = s4;s4 = peg$FAILED;
          }while (s4 !== peg$FAILED) {
            s3.push(s4);s4 = peg$currPos;s5 = peg$parsespace();if (s5 !== peg$FAILED) {
              s6 = peg$parseClause();if (s6 !== peg$FAILED) {
                peg$savedPos = s4;s5 = peg$c2(s2, s6);s4 = s5;
              } else {
                peg$currPos = s4;s4 = peg$FAILED;
              }
            } else {
              peg$currPos = s4;s4 = peg$FAILED;
            }
          }if (s3 !== peg$FAILED) {
            s4 = peg$parsespace();if (s4 === peg$FAILED) {
              s4 = null;
            }if (s4 !== peg$FAILED) {
              peg$savedPos = s0;s1 = peg$c3(s2, s3);s0 = s1;
            } else {
              peg$currPos = s0;s0 = peg$FAILED;
            }
          } else {
            peg$currPos = s0;s0 = peg$FAILED;
          }
        } else {
          peg$currPos = s0;s0 = peg$FAILED;
        }
      } else {
        peg$currPos = s0;s0 = peg$FAILED;
      }return s0;
    }function peg$parseClause() {
      var s0;s0 = peg$parseIsClause();if (s0 === peg$FAILED) {
        s0 = peg$parseFieldClause();if (s0 === peg$FAILED) {
          s0 = peg$parseTermClause();
        }
      }return s0;
    }function peg$parseTermClause() {
      var s0, s1, s2, s3;s0 = peg$currPos;s1 = peg$parsespace();if (s1 === peg$FAILED) {
        s1 = null;
      }if (s1 !== peg$FAILED) {
        if (input.charCodeAt(peg$currPos) === 45) {
          s2 = peg$c4;peg$currPos++;
        } else {
          s2 = peg$FAILED;if (peg$silentFails === 0) {
            peg$fail(peg$c5);
          }
        }if (s2 !== peg$FAILED) {
          s3 = peg$parsetermValue();if (s3 !== peg$FAILED) {
            peg$savedPos = s0;s1 = peg$c6(s3);s0 = s1;
          } else {
            peg$currPos = s0;s0 = peg$FAILED;
          }
        } else {
          peg$currPos = s0;s0 = peg$FAILED;
        }
      } else {
        peg$currPos = s0;s0 = peg$FAILED;
      }if (s0 === peg$FAILED) {
        s0 = peg$currPos;s1 = peg$parsespace();if (s1 === peg$FAILED) {
          s1 = null;
        }if (s1 !== peg$FAILED) {
          s2 = peg$parsetermValue();if (s2 !== peg$FAILED) {
            peg$savedPos = s0;s1 = peg$c7(s2);s0 = s1;
          } else {
            peg$currPos = s0;s0 = peg$FAILED;
          }
        } else {
          peg$currPos = s0;s0 = peg$FAILED;
        }
      }return s0;
    }function peg$parseIsClause() {
      var s0, s1, s2, s3;s0 = peg$currPos;s1 = peg$parsespace();if (s1 === peg$FAILED) {
        s1 = null;
      }if (s1 !== peg$FAILED) {
        if (input.charCodeAt(peg$currPos) === 45) {
          s2 = peg$c4;peg$currPos++;
        } else {
          s2 = peg$FAILED;if (peg$silentFails === 0) {
            peg$fail(peg$c5);
          }
        }if (s2 !== peg$FAILED) {
          s3 = peg$parseIsValue();if (s3 !== peg$FAILED) {
            peg$savedPos = s0;s1 = peg$c8(s3);s0 = s1;
          } else {
            peg$currPos = s0;s0 = peg$FAILED;
          }
        } else {
          peg$currPos = s0;s0 = peg$FAILED;
        }
      } else {
        peg$currPos = s0;s0 = peg$FAILED;
      }if (s0 === peg$FAILED) {
        s0 = peg$currPos;s1 = peg$parsespace();if (s1 === peg$FAILED) {
          s1 = null;
        }if (s1 !== peg$FAILED) {
          s2 = peg$parseIsValue();if (s2 !== peg$FAILED) {
            peg$savedPos = s0;s1 = peg$c9(s2);s0 = s1;
          } else {
            peg$currPos = s0;s0 = peg$FAILED;
          }
        } else {
          peg$currPos = s0;s0 = peg$FAILED;
        }
      }return s0;
    }function peg$parseIsValue() {
      var s0, s1, s2;s0 = peg$currPos;if (input.substr(peg$currPos, 3) === peg$c10) {
        s1 = peg$c10;peg$currPos += 3;
      } else {
        s1 = peg$FAILED;if (peg$silentFails === 0) {
          peg$fail(peg$c11);
        }
      }if (s1 !== peg$FAILED) {
        s2 = peg$parsevalue();if (s2 !== peg$FAILED) {
          peg$savedPos = s0;s1 = peg$c12(s2);s0 = s1;
        } else {
          peg$currPos = s0;s0 = peg$FAILED;
        }
      } else {
        peg$currPos = s0;s0 = peg$FAILED;
      }return s0;
    }function peg$parseFieldClause() {
      var s0, s1, s2, s3;s0 = peg$currPos;s1 = peg$parsespace();if (s1 === peg$FAILED) {
        s1 = null;
      }if (s1 !== peg$FAILED) {
        if (input.charCodeAt(peg$currPos) === 45) {
          s2 = peg$c4;peg$currPos++;
        } else {
          s2 = peg$FAILED;if (peg$silentFails === 0) {
            peg$fail(peg$c5);
          }
        }if (s2 !== peg$FAILED) {
          s3 = peg$parseFieldAndValue();if (s3 !== peg$FAILED) {
            peg$savedPos = s0;s1 = peg$c13(s3);s0 = s1;
          } else {
            peg$currPos = s0;s0 = peg$FAILED;
          }
        } else {
          peg$currPos = s0;s0 = peg$FAILED;
        }
      } else {
        peg$currPos = s0;s0 = peg$FAILED;
      }if (s0 === peg$FAILED) {
        s0 = peg$currPos;s1 = peg$parsespace();if (s1 === peg$FAILED) {
          s1 = null;
        }if (s1 !== peg$FAILED) {
          s2 = peg$parseFieldAndValue();if (s2 !== peg$FAILED) {
            peg$savedPos = s0;s1 = peg$c14(s2);s0 = s1;
          } else {
            peg$currPos = s0;s0 = peg$FAILED;
          }
        } else {
          peg$currPos = s0;s0 = peg$FAILED;
        }
      }return s0;
    }function peg$parseFieldAndValue() {
      var s0, s1, s2, s3;s0 = peg$currPos;s1 = peg$parsefieldName();if (s1 !== peg$FAILED) {
        if (input.charCodeAt(peg$currPos) === 58) {
          s2 = peg$c15;peg$currPos++;
        } else {
          s2 = peg$FAILED;if (peg$silentFails === 0) {
            peg$fail(peg$c16);
          }
        }if (s2 !== peg$FAILED) {
          s3 = peg$parsefieldValue();if (s3 !== peg$FAILED) {
            peg$savedPos = s0;s1 = peg$c17(s1, s3);s0 = s1;
          } else {
            peg$currPos = s0;s0 = peg$FAILED;
          }
        } else {
          peg$currPos = s0;s0 = peg$FAILED;
        }
      } else {
        peg$currPos = s0;s0 = peg$FAILED;
      }return s0;
    }function peg$parsefieldName() {
      var s0, s1, s2;peg$silentFails++;s0 = peg$currPos;s1 = [];s2 = peg$parsefieldChar();if (s2 !== peg$FAILED) {
        while (s2 !== peg$FAILED) {
          s1.push(s2);s2 = peg$parsefieldChar();
        }
      } else {
        s1 = peg$FAILED;
      }if (s1 !== peg$FAILED) {
        peg$savedPos = s0;s1 = peg$c19();
      }s0 = s1;peg$silentFails--;if (s0 === peg$FAILED) {
        s1 = peg$FAILED;if (peg$silentFails === 0) {
          peg$fail(peg$c18);
        }
      }return s0;
    }function peg$parsefieldChar() {
      var s0;s0 = peg$parsealnum();if (s0 === peg$FAILED) {
        s0 = peg$parseescapedChar();
      }return s0;
    }function peg$parsefieldValue() {
      var s0, s1;peg$silentFails++;s0 = peg$parsefieldValues();if (s0 === peg$FAILED) {
        s0 = peg$parsevalue();
      }peg$silentFails--;if (s0 === peg$FAILED) {
        s1 = peg$FAILED;if (peg$silentFails === 0) {
          peg$fail(peg$c20);
        }
      }return s0;
    }function peg$parsefieldValues() {
      var s0, s1, s2, s3, s4, s5, s6, s7, s8, s9;s0 = peg$currPos;if (input.charCodeAt(peg$currPos) === 40) {
        s1 = peg$c21;peg$currPos++;
      } else {
        s1 = peg$FAILED;if (peg$silentFails === 0) {
          peg$fail(peg$c22);
        }
      }if (s1 !== peg$FAILED) {
        s2 = peg$parsespace();if (s2 === peg$FAILED) {
          s2 = null;
        }if (s2 !== peg$FAILED) {
          s3 = peg$parsevalue();if (s3 !== peg$FAILED) {
            s4 = [];s5 = peg$currPos;s6 = peg$parsespace();if (s6 !== peg$FAILED) {
              s7 = peg$currPos;if (peg$c23.test(input.charAt(peg$currPos))) {
                s8 = input.charAt(peg$currPos);peg$currPos++;
              } else {
                s8 = peg$FAILED;if (peg$silentFails === 0) {
                  peg$fail(peg$c24);
                }
              }if (s8 !== peg$FAILED) {
                if (peg$c25.test(input.charAt(peg$currPos))) {
                  s9 = input.charAt(peg$currPos);peg$currPos++;
                } else {
                  s9 = peg$FAILED;if (peg$silentFails === 0) {
                    peg$fail(peg$c26);
                  }
                }if (s9 !== peg$FAILED) {
                  s8 = [s8, s9];s7 = s8;
                } else {
                  peg$currPos = s7;s7 = peg$FAILED;
                }
              } else {
                peg$currPos = s7;s7 = peg$FAILED;
              }if (s7 !== peg$FAILED) {
                s8 = peg$parsespace();if (s8 !== peg$FAILED) {
                  s9 = peg$parsevalue();if (s9 !== peg$FAILED) {
                    peg$savedPos = s5;s6 = peg$c27(s3, s9);s5 = s6;
                  } else {
                    peg$currPos = s5;s5 = peg$FAILED;
                  }
                } else {
                  peg$currPos = s5;s5 = peg$FAILED;
                }
              } else {
                peg$currPos = s5;s5 = peg$FAILED;
              }
            } else {
              peg$currPos = s5;s5 = peg$FAILED;
            }while (s5 !== peg$FAILED) {
              s4.push(s5);s5 = peg$currPos;s6 = peg$parsespace();if (s6 !== peg$FAILED) {
                s7 = peg$currPos;if (peg$c23.test(input.charAt(peg$currPos))) {
                  s8 = input.charAt(peg$currPos);peg$currPos++;
                } else {
                  s8 = peg$FAILED;if (peg$silentFails === 0) {
                    peg$fail(peg$c24);
                  }
                }if (s8 !== peg$FAILED) {
                  if (peg$c25.test(input.charAt(peg$currPos))) {
                    s9 = input.charAt(peg$currPos);peg$currPos++;
                  } else {
                    s9 = peg$FAILED;if (peg$silentFails === 0) {
                      peg$fail(peg$c26);
                    }
                  }if (s9 !== peg$FAILED) {
                    s8 = [s8, s9];s7 = s8;
                  } else {
                    peg$currPos = s7;s7 = peg$FAILED;
                  }
                } else {
                  peg$currPos = s7;s7 = peg$FAILED;
                }if (s7 !== peg$FAILED) {
                  s8 = peg$parsespace();if (s8 !== peg$FAILED) {
                    s9 = peg$parsevalue();if (s9 !== peg$FAILED) {
                      peg$savedPos = s5;s6 = peg$c27(s3, s9);s5 = s6;
                    } else {
                      peg$currPos = s5;s5 = peg$FAILED;
                    }
                  } else {
                    peg$currPos = s5;s5 = peg$FAILED;
                  }
                } else {
                  peg$currPos = s5;s5 = peg$FAILED;
                }
              } else {
                peg$currPos = s5;s5 = peg$FAILED;
              }
            }if (s4 !== peg$FAILED) {
              s5 = peg$parsespace();if (s5 === peg$FAILED) {
                s5 = null;
              }if (s5 !== peg$FAILED) {
                if (input.charCodeAt(peg$currPos) === 41) {
                  s6 = peg$c28;peg$currPos++;
                } else {
                  s6 = peg$FAILED;if (peg$silentFails === 0) {
                    peg$fail(peg$c29);
                  }
                }if (s6 !== peg$FAILED) {
                  peg$savedPos = s0;s1 = peg$c30(s3, s4);s0 = s1;
                } else {
                  peg$currPos = s0;s0 = peg$FAILED;
                }
              } else {
                peg$currPos = s0;s0 = peg$FAILED;
              }
            } else {
              peg$currPos = s0;s0 = peg$FAILED;
            }
          } else {
            peg$currPos = s0;s0 = peg$FAILED;
          }
        } else {
          peg$currPos = s0;s0 = peg$FAILED;
        }
      } else {
        peg$currPos = s0;s0 = peg$FAILED;
      }return s0;
    }function peg$parsetermValue() {
      var s0, s1;peg$silentFails++;s0 = peg$parsevalue();peg$silentFails--;if (s0 === peg$FAILED) {
        s1 = peg$FAILED;if (peg$silentFails === 0) {
          peg$fail(peg$c31);
        }
      }return s0;
    }function peg$parsevalue() {
      var s0, s1, s2, s3, s4, s5;s0 = peg$parseword();if (s0 === peg$FAILED) {
        s0 = peg$currPos;if (input.charCodeAt(peg$currPos) === 34) {
          s1 = peg$c32;peg$currPos++;
        } else {
          s1 = peg$FAILED;if (peg$silentFails === 0) {
            peg$fail(peg$c33);
          }
        }if (s1 !== peg$FAILED) {
          s2 = peg$parsespace();if (s2 === peg$FAILED) {
            s2 = null;
          }if (s2 !== peg$FAILED) {
            s3 = peg$parsephrase();if (s3 !== peg$FAILED) {
              s4 = peg$parsespace();if (s4 === peg$FAILED) {
                s4 = null;
              }if (s4 !== peg$FAILED) {
                if (input.charCodeAt(peg$currPos) === 34) {
                  s5 = peg$c32;peg$currPos++;
                } else {
                  s5 = peg$FAILED;if (peg$silentFails === 0) {
                    peg$fail(peg$c33);
                  }
                }if (s5 !== peg$FAILED) {
                  peg$savedPos = s0;s1 = peg$c34(s3);s0 = s1;
                } else {
                  peg$currPos = s0;s0 = peg$FAILED;
                }
              } else {
                peg$currPos = s0;s0 = peg$FAILED;
              }
            } else {
              peg$currPos = s0;s0 = peg$FAILED;
            }
          } else {
            peg$currPos = s0;s0 = peg$FAILED;
          }
        } else {
          peg$currPos = s0;s0 = peg$FAILED;
        }
      }return s0;
    }function peg$parsephrase() {
      var s0, s1, s2, s3, s4, s5;s0 = peg$currPos;s1 = peg$parseword();if (s1 !== peg$FAILED) {
        s2 = [];s3 = peg$currPos;s4 = peg$parsespace();if (s4 !== peg$FAILED) {
          s5 = peg$parseword();if (s5 !== peg$FAILED) {
            s4 = [s4, s5];s3 = s4;
          } else {
            peg$currPos = s3;s3 = peg$FAILED;
          }
        } else {
          peg$currPos = s3;s3 = peg$FAILED;
        }while (s3 !== peg$FAILED) {
          s2.push(s3);s3 = peg$currPos;s4 = peg$parsespace();if (s4 !== peg$FAILED) {
            s5 = peg$parseword();if (s5 !== peg$FAILED) {
              s4 = [s4, s5];s3 = s4;
            } else {
              peg$currPos = s3;s3 = peg$FAILED;
            }
          } else {
            peg$currPos = s3;s3 = peg$FAILED;
          }
        }if (s2 !== peg$FAILED) {
          peg$savedPos = s0;s1 = peg$c19();s0 = s1;
        } else {
          peg$currPos = s0;s0 = peg$FAILED;
        }
      } else {
        peg$currPos = s0;s0 = peg$FAILED;
      }return s0;
    }function peg$parseword() {
      var s0, s1, s2;s0 = peg$currPos;s1 = [];s2 = peg$parsevalueChar();if (s2 !== peg$FAILED) {
        while (s2 !== peg$FAILED) {
          s1.push(s2);s2 = peg$parsevalueChar();
        }
      } else {
        s1 = peg$FAILED;
      }if (s1 !== peg$FAILED) {
        peg$savedPos = s0;s1 = peg$c19();
      }s0 = s1;return s0;
    }function peg$parsevalueChar() {
      var s0;s0 = peg$parsealnum();if (s0 === peg$FAILED) {
        s0 = peg$parseescapedChar();
      }return s0;
    }function peg$parseescapedChar() {
      var s0, s1, s2;s0 = peg$currPos;if (input.charCodeAt(peg$currPos) === 92) {
        s1 = peg$c35;peg$currPos++;
      } else {
        s1 = peg$FAILED;if (peg$silentFails === 0) {
          peg$fail(peg$c36);
        }
      }if (s1 !== peg$FAILED) {
        s2 = peg$parsereservedChar();if (s2 !== peg$FAILED) {
          s1 = [s1, s2];s0 = s1;
        } else {
          peg$currPos = s0;s0 = peg$FAILED;
        }
      } else {
        peg$currPos = s0;s0 = peg$FAILED;
      }return s0;
    }function peg$parsereservedChar() {
      var s0;if (peg$c37.test(input.charAt(peg$currPos))) {
        s0 = input.charAt(peg$currPos);peg$currPos++;
      } else {
        s0 = peg$FAILED;if (peg$silentFails === 0) {
          peg$fail(peg$c38);
        }
      }return s0;
    }function peg$parsealnum() {
      var s0, s1;peg$silentFails++;if (peg$c40.test(input.charAt(peg$currPos))) {
        s0 = input.charAt(peg$currPos);peg$currPos++;
      } else {
        s0 = peg$FAILED;if (peg$silentFails === 0) {
          peg$fail(peg$c41);
        }
      }peg$silentFails--;if (s0 === peg$FAILED) {
        s1 = peg$FAILED;if (peg$silentFails === 0) {
          peg$fail(peg$c39);
        }
      }return s0;
    }function peg$parsespace() {
      var s0, s1;peg$silentFails++;s0 = [];if (peg$c43.test(input.charAt(peg$currPos))) {
        s1 = input.charAt(peg$currPos);peg$currPos++;
      } else {
        s1 = peg$FAILED;if (peg$silentFails === 0) {
          peg$fail(peg$c44);
        }
      }if (s1 !== peg$FAILED) {
        while (s1 !== peg$FAILED) {
          s0.push(s1);if (peg$c43.test(input.charAt(peg$currPos))) {
            s1 = input.charAt(peg$currPos);peg$currPos++;
          } else {
            s1 = peg$FAILED;if (peg$silentFails === 0) {
              peg$fail(peg$c44);
            }
          }
        }
      } else {
        s0 = peg$FAILED;
      }peg$silentFails--;if (s0 === peg$FAILED) {
        s1 = peg$FAILED;if (peg$silentFails === 0) {
          peg$fail(peg$c42);
        }
      }return s0;
    }var _options = options,
        AST = _options.AST,
        unescapeValue = _options.unescapeValue;
    peg$result = peg$startRuleFunction();if (peg$result !== peg$FAILED && peg$currPos === input.length) {
      return peg$result;
    } else {
      if (peg$result !== peg$FAILED && peg$currPos < input.length) {
        peg$fail(peg$endExpectation());
      }throw peg$buildStructuredError(peg$maxFailExpected, peg$maxFailPos < input.length ? input.charAt(peg$maxFailPos) : null, peg$maxFailPos < input.length ? peg$computeLocation(peg$maxFailPos, peg$maxFailPos + 1) : peg$computeLocation(peg$maxFailPos, peg$maxFailPos));
    }
  }return { SyntaxError: peg$SyntaxError, parse: peg$parse };
}();

var printValue = function printValue(value) {
  if (value.match(/\s/)) {
    return '"' + escapeValue(value) + '"';
  }
  return escapeValue(value);
};

var defaultSyntax = exports.defaultSyntax = Object.freeze({

  parse: function parse(query) {
    var clauses = parser.parse(query, { AST: _ast.AST, unescapeValue: unescapeValue });
    return _ast.AST.create(clauses);
  },

  print: function print(ast) {
    return ast.clauses.reduce(function (text, clause) {
      var prefix = _ast.AST.Match.isMustClause(clause) ? '' : '-';
      switch (clause.type) {
        case _ast.AST.Field.TYPE:
          if ((0, _predicate.isArray)(clause.value)) {
            return text + ' ' + prefix + escapeValue(clause.field) + ':(' + clause.value.map(function (val) {
              return printValue(val);
            }).join(' or ') + ')';
          }
          return text + ' ' + prefix + escapeValue(clause.field) + ':' + printValue(clause.value);
        case _ast.AST.Is.TYPE:
          return text + ' ' + prefix + 'is:' + escapeValue(clause.flag);
        case _ast.AST.Term.TYPE:
          return text + ' ' + prefix + printValue(clause.value);
        default:
          return text;
      }
    }, '').trim();
  }

});
