// @elastic/eui/lib/components/header/header_logo.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.EuiHeaderLogo = void 0;

require("core-js/modules/es6.object.assign");

require("core-js/modules/web.dom.iterable");

require("core-js/modules/es6.array.iterator");

require("core-js/modules/es6.object.keys");

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _classnames = _interopRequireDefault(require("classnames"));

var _icon = require("../icon");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

var EuiHeaderLogo = function EuiHeaderLogo(_ref) {
  var iconType = _ref.iconType,
      iconTitle = _ref.iconTitle,
      href = _ref.href,
      children = _ref.children,
      className = _ref.className,
      rest = _objectWithoutProperties(_ref, ["iconType", "iconTitle", "href", "children", "className"]);

  var classes = (0, _classnames.default)('euiHeaderLogo', className);
  return _react.default.createElement("a", _extends({
    href: href,
    className: classes
  }, rest), _react.default.createElement(_icon.EuiIcon, {
    className: "euiHeaderLogo__icon",
    size: "l",
    type: iconType,
    title: iconTitle
  }), children && _react.default.createElement("span", {
    className: "euiHeaderLogo__text"
  }, children));
};

exports.EuiHeaderLogo = EuiHeaderLogo;
EuiHeaderLogo.propTypes = {
  href: _propTypes.default.string,
  children: _propTypes.default.node,
  iconType: _propTypes.default.string,
  iconTitle: _propTypes.default.string
};
EuiHeaderLogo.defaultProps = {
  iconType: 'logoElastic',
  iconTitle: 'Elastic'
};
EuiHeaderLogo.__docgenInfo = {
  "description": "",
  "methods": [],
  "displayName": "EuiHeaderLogo",
  "props": {
    "iconType": {
      "defaultValue": {
        "value": "'logoElastic'",
        "computed": false
      },
      "type": {
        "name": "string"
      },
      "required": false,
      "description": ""
    },
    "iconTitle": {
      "defaultValue": {
        "value": "'Elastic'",
        "computed": false
      },
      "type": {
        "name": "string"
      },
      "required": false,
      "description": ""
    },
    "href": {
      "type": {
        "name": "string"
      },
      "required": false,
      "description": ""
    },
    "children": {
      "type": {
        "name": "node"
      },
      "required": false,
      "description": ""
    }
  }
};
