// @elastic/eui/lib/components/form/select/select.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.EuiSelect = undefined;

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _form_control_layout = require('../form_control_layout');

var _validatable_control = require('../validatable_control');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

var EuiSelect = function EuiSelect(_ref) {
  var className = _ref.className,
      options = _ref.options,
      id = _ref.id,
      name = _ref.name,
      inputRef = _ref.inputRef,
      isInvalid = _ref.isInvalid,
      fullWidth = _ref.fullWidth,
      isLoading = _ref.isLoading,
      hasNoInitialSelection = _ref.hasNoInitialSelection,
      rest = _objectWithoutProperties(_ref, ['className', 'options', 'id', 'name', 'inputRef', 'isInvalid', 'fullWidth', 'isLoading', 'hasNoInitialSelection']);

  var classes = (0, _classnames2.default)('euiSelect', {
    'euiSelect--fullWidth': fullWidth,
    'euiSelect-isLoading': isLoading
  }, className);

  var emtpyOptionNode = void 0;
  if (hasNoInitialSelection) {
    emtpyOptionNode = _react2.default.createElement(
      'option',
      { selected: true, disabled: true, hidden: true, style: { display: 'none' } },
      '\xA0'
    );
  }

  return _react2.default.createElement(
    _form_control_layout.EuiFormControlLayout,
    {
      icon: 'arrowDown',
      iconSide: 'right',
      fullWidth: fullWidth,
      isLoading: isLoading
    },
    _react2.default.createElement(
      _validatable_control.EuiValidatableControl,
      { isInvalid: isInvalid },
      _react2.default.createElement(
        'select',
        _extends({
          id: id,
          name: name,
          className: classes,
          ref: inputRef
        }, rest),
        emtpyOptionNode,
        options.map(function (option, index) {
          var text = option.text,
              rest = _objectWithoutProperties(option, ['text']);

          return _react2.default.createElement(
            'option',
            _extends({}, rest, { key: index }),
            text
          );
        })
      )
    )
  );
};

exports.EuiSelect = EuiSelect;
EuiSelect.propTypes = {
  name: _propTypes2.default.string,
  id: _propTypes2.default.string,
  options: _propTypes2.default.arrayOf(_propTypes2.default.shape({
    text: _propTypes2.default.string.isRequired
  })).isRequired,
  isInvalid: _propTypes2.default.bool,
  fullWidth: _propTypes2.default.bool,
  isLoading: _propTypes2.default.bool,

  /**
   * Simulates no selection by creating an empty, selected, hidden first option
   */
  hasNoInitialSelection: _propTypes2.default.bool,
  inputRef: _propTypes2.default.func
};

EuiSelect.defaultProps = {
  options: [],
  fullWidth: false,
  isLoading: false,
  hasNoInitialSelection: false
};
EuiSelect.__docgenInfo = [{
  'description': '',
  'methods': [],
  'props': {
    'name': {
      'type': {
        'name': 'string'
      },
      'required': false,
      'description': ''
    },
    'id': {
      'type': {
        'name': 'string'
      },
      'required': false,
      'description': ''
    },
    'options': {
      'type': {
        'name': 'arrayOf',
        'value': {
          'name': 'shape',
          'value': {
            'text': {
              'name': 'string',
              'required': true
            }
          }
        }
      },
      'required': true,
      'description': '',
      'defaultValue': {
        'value': '[]',
        'computed': false
      }
    },
    'isInvalid': {
      'type': {
        'name': 'bool'
      },
      'required': false,
      'description': ''
    },
    'fullWidth': {
      'type': {
        'name': 'bool'
      },
      'required': false,
      'description': '',
      'defaultValue': {
        'value': 'false',
        'computed': false
      }
    },
    'isLoading': {
      'type': {
        'name': 'bool'
      },
      'required': false,
      'description': '',
      'defaultValue': {
        'value': 'false',
        'computed': false
      }
    },
    'hasNoInitialSelection': {
      'type': {
        'name': 'bool'
      },
      'required': false,
      'description': 'Simulates no selection by creating an empty, selected, hidden first option',
      'defaultValue': {
        'value': 'false',
        'computed': false
      }
    },
    'inputRef': {
      'type': {
        'name': 'func'
      },
      'required': false,
      'description': ''
    }
  }
}];
