// @elastic/eui/es/components/loading/loading_chart.js
import "core-js/modules/es6.object.assign";
import "core-js/modules/web.dom.iterable";
import "core-js/modules/es6.array.iterator";
import "core-js/modules/es6.object.keys";

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
var sizeToClassNameMap = {
  m: 'euiLoadingChart--medium',
  l: 'euiLoadingChart--large',
  xl: 'euiLoadingChart--xLarge'
};
export var SIZES = Object.keys(sizeToClassNameMap);
export var EuiLoadingChart = function EuiLoadingChart(_ref) {
  var size = _ref.size,
      mono = _ref.mono,
      className = _ref.className,
      rest = _objectWithoutProperties(_ref, ["size", "mono", "className"]);

  var classes = classNames('euiLoadingChart', mono === true ? 'euiLoadingChart--mono' : '', className, sizeToClassNameMap[size]);
  return React.createElement("div", _extends({
    className: classes
  }, rest), React.createElement("div", {
    className: "euiLoadingChart__bar"
  }), React.createElement("div", {
    className: "euiLoadingChart__bar"
  }), React.createElement("div", {
    className: "euiLoadingChart__bar"
  }), React.createElement("div", {
    className: "euiLoadingChart__bar"
  }));
};
EuiLoadingChart.propTypes = {
  mono: PropTypes.bool,
  size: PropTypes.oneOf(SIZES)
};
EuiLoadingChart.defaultProps = {
  mono: false
};
EuiLoadingChart.__docgenInfo = {
  "description": "",
  "methods": [],
  "displayName": "EuiLoadingChart",
  "props": {
    "mono": {
      "defaultValue": {
        "value": "false",
        "computed": false
      },
      "type": {
        "name": "bool"
      },
      "required": false,
      "description": ""
    },
    "size": {
      "type": {
        "name": "enum",
        "value": [{
          "value": "\"m\"",
          "computed": false
        }, {
          "value": "\"l\"",
          "computed": false
        }, {
          "value": "\"xl\"",
          "computed": false
        }]
      },
      "required": false,
      "description": ""
    }
  }
};
