// @elastic/eui/es/components/text/text_align.js
import "core-js/modules/es6.object.assign";
import "core-js/modules/web.dom.iterable";
import "core-js/modules/es6.array.iterator";
import "core-js/modules/es6.object.keys";

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
export var alignmentToClassNameMap = {
  'left': 'euiTextAlign--left',
  'right': 'euiTextAlign--right',
  'center': 'euiTextAlign--center'
};
export var ALIGNMENTS = Object.keys(alignmentToClassNameMap);
export var EuiTextAlign = function EuiTextAlign(_ref) {
  var children = _ref.children,
      className = _ref.className,
      textAlign = _ref.textAlign,
      rest = _objectWithoutProperties(_ref, ["children", "className", "textAlign"]);

  var classes = classNames('euiTextAlign', alignmentToClassNameMap[textAlign], className);
  return React.createElement("div", _extends({
    className: classes
  }, rest), children);
};
EuiTextAlign.propTypes = {
  children: PropTypes.node,
  className: PropTypes.string,
  textAlign: PropTypes.oneOf(ALIGNMENTS)
};
EuiTextAlign.defaultProps = {
  textAlign: 'left'
};
EuiTextAlign.__docgenInfo = {
  "description": "",
  "methods": [],
  "displayName": "EuiTextAlign",
  "props": {
    "textAlign": {
      "defaultValue": {
        "value": "'left'",
        "computed": false
      },
      "type": {
        "name": "enum",
        "value": [{
          "value": "\"left\"",
          "computed": false
        }, {
          "value": "\"right\"",
          "computed": false
        }, {
          "value": "\"center\"",
          "computed": false
        }]
      },
      "required": false,
      "description": ""
    },
    "children": {
      "type": {
        "name": "node"
      },
      "required": false,
      "description": ""
    },
    "className": {
      "type": {
        "name": "string"
      },
      "required": false,
      "description": ""
    }
  }
};
