// @elastic/eui/lib/components/table_of_records/expanded_record_actions.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ExpandedRecordActions = undefined;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _default_record_action = require('./default_record_action');

var _custom_record_action = require('./custom_record_action');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var ExpandedRecordActions = exports.ExpandedRecordActions = function ExpandedRecordActions(_ref) {
  var actions = _ref.actions,
      visible = _ref.visible,
      recordId = _ref.recordId,
      record = _ref.record,
      model = _ref.model,
      actionEnabled = _ref.actionEnabled;


  return actions.reduce(function (tools, action, index) {
    var available = action.available ? action.available(record, model) : true;
    if (!available) {
      return tools;
    }
    var enabled = actionEnabled(action);
    var key = 'record_action_' + recordId + '_' + index;
    if (action.render) {
      // custom action has a render function
      tools.push(_react2.default.createElement(_custom_record_action.CustomRecordAction, {
        key: key,
        index: index,
        action: action,
        enabled: enabled,
        visible: visible,
        recordId: recordId,
        record: record,
        model: model
      }));
    } else {
      tools.push(_react2.default.createElement(_default_record_action.DefaultRecordAction, {
        key: key,
        index: index,
        action: action,
        enabled: enabled,
        visible: visible,
        recordId: recordId,
        record: record,
        model: model
      }));
    }
    return tools;
  }, []);
};
